sklearn/utils/stats.py

```diff
@@ -40,33 +49,42 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     # When sample_weight 1D, repeat for each array.shape[1]
     if array.shape != sample_weight.shape and array.shape[0] == sample_weight.shape[0]:
         sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
+
+    # Sort `array` and `sample_weight` along axis=0; set `sample_weight` for nan input
+    # to 0:
     sorted_idx = np.argsort(array, axis=0)
     sorted_weights = np.take_along_axis(sample_weight, sorted_idx, axis=0)
+    sorted_nan_mask = np.take_along_axis(np.isnan(array), sorted_idx, axis=0)
+    sorted_weights[sorted_nan_mask] = 0
 
-    # Find index of median prediction for each sample
+    # Compute the weighted cumulative distribution function (CDF) based on
+    # `sample_weight` and scale `percentile` along it:
     weight_cdf = stable_cumsum(sorted_weights, axis=0)
-    adjusted_percentile = percentile / 100 * weight_cdf[-1]
+    adjusted_percentile = percentile_rank / 100 * weight_cdf[-1]
 
-    # For percentile=0, ignore leading observations with sample_weight=0. GH20528
+    # For percentile_rank=0, ignore leading observations with sample_weight=0; see PR
+    # #20528
     mask = adjusted_percentile == 0
     adjusted_percentile[mask] = np.nextafter(
         adjusted_percentile[mask], adjusted_percentile[mask] + 1
     )
 
+    # Find index `adjusted_percentile` would have in `weight_cdf`:
```

Comment:
I find the following even clearer:

```suggestion
    # For each column index (j), find the row index (i), such that:
    # weight_cdf[i-1, j] < adjusted_percentile[j] <= weight_cdf[i, j]
```

Reply:
I feel to only use this inequality as a comment would add to the complexity of this function rather than to reduce it.

Don't get me wrong, but can we still add some words that describe the purpose of this code instead of only describing how it's done? I bet it's not only unclear to me.

Maybe like this:
```
    # For each column index (j), find the row index (i), such that: 
    # weight_cdf[i-1, j] < adjusted_percentile[j] <= weight_cdf[i, j]; this means to
    # find the indices `adjusted_percentile_rank` would have in `weight_cdf`:
```